### DISCUSSION TITLE
greenlet_spawn has not been called; can't call await_only() here

### DISCUSSION THREAD

**Discussion by cnicodeme** [sqlalchemy/sqlalchemy#9757] [Usage Questions]
Hi team!

I'm having this issue from time to time. From my understanding, the problems is related to SQLAlchemy (and not the underlying asyncmy library I'm using, but maybe I'm missing something).

Here's the error:

> (sqlalchemy.exc.MissingGreenlet) greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place?

And the stacktrace:

```
(Background on this error at: https://sqlalche.me/e/20/xd2s)
  File "/var/www/getfernand.com/www/commands/retry_pendings.py", line 60, in run
    await process_incoming_email_sub(
  File "/var/www/getfernand.com/www/apps/webhooks/tasks.py", line 247, in process_incoming_email_sub
    organization_id = domain.organization_id
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py", line 557, in __get__
    return self.impl.get(state, dict_)  # type: ignore[no-any-return]
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py", line 1070, in get
    value = self._fire_loader_callables(state, key, passive)
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py", line 1100, in _fire_loader_callables
    return state._load_expired(state, passive)
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/state.py", line 807, in _load_expired
    self.manager.expired_attribute_loader(self, toload, passive)
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 1626, in load_scalar_attributes
    result = load_on_ident(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 482, in load_on_ident
    return load_on_pk_identity(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/loading.py", line 668, in load_on_pk_identity
    session.execute(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 2229, in execute
    return self._execute_internal(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 2124, in _execute_internal
    result: Result[Any] = compile_state_cls.orm_execute_statement(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/orm/context.py", line 253, in orm_execute_statement
    result = conn.execute(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1414, in execute
    return meth(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 486, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1638, in _execute_clauseelement
    ret = self._execute_context(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1814, in _execute_context
    self._handle_dbapi_exception(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2326, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1808, in _execute_context
    context = constructor(
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1203, in _init_compiled
    self.cursor = self.create_cursor()
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1530, in create_cursor
    return self.create_default_cursor()
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1533, in create_default_cursor
    return self._dbapi_connection.cursor()
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 1494, in cursor
    return self.dbapi_connection.cursor(*args, **kwargs)
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/asyncmy.py", line 215, in cursor
    return AsyncAdapt_asyncmy_cursor(self)
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/asyncmy.py", line 61, in __init__
    self._cursor = self.await_(cursor.__aenter__())
  File "/var/www/getfernand.com/env/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 93, in await_only
    raise exc.MissingGreenlet(
```

Happy to provide any other insights if you need :)

**Comment by cnicodeme** [comment]
I initially thought the issue was because the call was made after the async loop was closed but it's not the case, it was running inside a while loop, when this occurred, and some code was executed asynchronously before, and after, that issue.

**Comment by zzzeek** [comment]
hi - 

this error message should include a hyperlink that will direct to this page:

https://docs.sqlalchemy.org/en/20/errors.html#missinggreenlet


A complete description of what this error means and how to mitigate follows from there.

**Comment by cnicodeme** [comment]
Yes the link was given and I followed it. But reading didn't helped me for me. From my understanding, it means the call was made outside of the ... greenlet spawn context. I'm understanding that this means it was called outside of the loop it was executed from ? (or is it something else completely?)

It also talks about lazy loading, which I could understand, but in this case, the line in question is to get a column in the model, not a foreign key.

I'm continuing to search for this in the hope to find something, but if I can get any help, I'm happy too

**Reply by zzzeek** [reply]
the stack trace refers to code you have that looks like this:

```py
def process_incoming_email_sub(...):
    organization_id = domain.organization_id
```

Above, the object represented by "domain" we would assume is expired.   The contents of the "organization_id" attribute are not present, and when you access it, the ORM refreshes the whole object based on its primary key.   "lazy loading" here is meant in the more general sense indicated in the [glossary](https://docs.sqlalchemy.org/en/20/glossary.html#term-lazy-loading) - from a SQLAlchemy ORM perspective, column-based lazy loads are actually "expired column loads" or "deferred column loads".

The overall issue of expired objects when dealing with the ORM, whether or not under asyncio, is normally because the `Session` or `AsyncSession` would benefit from setting [expire_on_commit](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.expire_on_commit) to False, which should be the case for all examples of ORM use in the asyncio document.

**Comment by cnicodeme** [comment]
Thank you! That makes so much more sense because I was definitely thinking that "lazy loading" was about relationships that I had removed because of the async nature of the code, so it didn't made sense.

I checked my code and the "expire_on_commit" property is currently set to False, so the configuration appears to correct, but somehow, the model was either expired or deferred loading.

Also, that column (I had a few other similar exception for other columns) are standard and aren't specifically asking for lazy loading:

```python
organization_id = Column(Integer, ForeignKey('organizations.id'), nullable=False)
```

Maybe it's implied by some parameter (ForeignKey?).

Is there a way I can either ensure that value isn't deferred/lazy loaded, or to run ... "something"? that would avoid this issue before calling 

```python
def process_incoming_email_sub(...):
    organization_id = domain.organization_id
```

(ideally, this would be generally done).

Thank you!

**Reply by zzzeek** [reply]
yes in the next SQLAlchemy release we have a helper for this called [AsyncAttrs](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs)

however it would be good to figure out why your "domain" object is expired.  is that a joined-inheritance model and that's a subclass?  that would also cause lazy loading depending on how domain was queried.

**Reply by CaselIT** [reply]
> Is there a way I can either ensure that value isn't deferred/lazy loaded, or to run ... "something"? that would avoid this issue before calling

You can use refresh before accessing `organization_id`: https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncSession.refresh

**Reply by cnicodeme** [reply]
I'm sorry I haven't answered sooner. I put that issue on hold hoping to get some time to get back to it sooner, but lost track of it...

I have the exact same issue but elsewhere in the code (oddly, same function) and same logic (assigning and from a retrieved object).

This time, the issue is occurring at this line:

   conversation.channel_id = channel.id

And the code that loads the channel is the following:

```python
    @classmethod
    async def find_by_email(cls, email, organization_id=None, include_removed=False) -> 'Channel':
        query = cls.query.filter(cls.email == email)
        if not include_removed:
            query = query.filter(cls.removed == None)  # noqa: E711

        if organization_id:
            query = query.join(Domain).filter(Domain.organization_id == organization_id)

        return await query.first()
```

There is definitely a join with Domain, to retrieve it from the proper organization.

Happy to provide more details.

Again, the "Domain" here is only linked to the "Channel" domain via a ForeignKey but no lazy loading or relationship since I'm on async.

In this case,  the Channel model has the following details:

```python
domain_id       = Column(Integer(unsigned=True), ForeignKey('domains.id'), nullable=False)
```

But that might help =>

When loading Channel, I have a function 'async def get_domain(self)` to retrieve the Domain instance from its id.
The function is as follow (in `Channel`)

```python
    @memoize_getter
    async def get_domain(self) -> Domain:
        return Domain.query.filter(Domain.id == self.domain_id)
```

Here's what `memoize_getter` does:

```python
def memoize_getter(f):
    @functools.wraps(f)
    async def decorator(instance, *args, **kwargs):
        property_name = '_' + f.__name__.replace('get_', '')
        dict_ = instance_dict(instance)

        if property_name not in dict_:
            dict_[property_name] = await await_if_needed(f, instance, *args, **kwargs)
            if isinstance(dict_[property_name], SelectWrapper):
                dict_[property_name] = await dict_[property_name].first()

        return dict_[property_name]

    return decorator
```

The purpose of this was to set up some kind of cache when retrieving the same data.

Now, following these issues, I'm suspecting this should be causing the issue (can you confirm?) but this leads me to another question:

Does SQLAlchemy handle some cache for me? (In this case, If I call the `get_domain` function three times in a row, will a call to the DB be made 3 times ? or just once and then returns a cached result until it needs to be re-loaded again )?

Thank you deeply for your help (not just for here, but for everytime).

**Reply by cnicodeme** [reply]
I was able to make a reproducible issue (I think?) by doing the following:

```
    c = await Channel.find_by_email('some@email.com')
    domain = await c.get_domain()
    db.write.expire(domain)
    print(domain.id)  # Fails here
```

I just don't know why my `domain` expires in the above function (`process_incoming_email_sub`)

My db object is wrapped in a aenter/aexit, and on the exit, calls `commit` and the `close` this session.

Maybe calling `commit` puts all the objects in expired mode ?

**Reply by zzzeek** [reply]
**edit::** this comment is irrelevant, that's not a Query object they are using

The issue with returning a `Domain.query` from a `@memoize_getter` is that this `Query` object will be bound to a fixed `Session` that would no longer be the `Session` in use; it would actually store this `Session` around which may even be linked to live connections so using a memoize like this is a bad idea.    you would at the very least unlink the session:

```
    @memoize_getter
    async def get_domain(self) -> Domain:
        q = Domain.query.filter(Domain.id == self.domain_id)
        q.session = None
        return q
```


if you use new-style queries you can return a `select()` construct from a memoized function without issue.

**Reply by cnicodeme** [reply]
Thank you. Can you point me to where I can implement the "new-style" queries?

**Reply by cnicodeme** [reply]
(Doing the `q.session = None` throws the error : 'SelectWrapper' object has no attribute 'session' )

**Reply by zzzeek** [reply]
> Doing the q.session = None throws the error : 'SelectWrapper' object has no attribute 'session'
 
This is some third party extension then, that's not a Query.  looked like a flask query.   you can disregard.    

you actually can't use Query with asyncio in any case so my comment is irrelevant.

I don't totally follow what your `@memoize_getter` is doing, somehow it's executing a statement it seems, but if you suspect it's causing problems, remove it and see if the problems go away.

>Maybe calling commit puts all the objects in expired mode ?

it does.   this is a key thing you want to disable when using asyncio, see the fifth bullet at https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession


> Does SQLAlchemy handle some cache for me? (In this case, If I call the get_domain function three times in a row, will a call to the DB be made 3 times ? or just once and then returns a cached result until it needs to be re-loaded again )?

SQLAlchemy always emits a database query when its told to, however if you are returning ORM objects they will be consulted from the local identity map first for the `Session` in use.   the tutorial at https://docs.sqlalchemy.org/en/20/tutorial/orm_data_manipulation.html#inserting-rows-using-the-orm-unit-of-work-pattern tries to walk through this concept

**Reply by cnicodeme** [reply]
Thank you for the clarifications.

The memoize getter receive either a Optional[_R] or a SelectWrapper.
When it's a SelectWrapper, the call to `.first()` is done inside the memoize function.

The idea is two fold: first to simplify the code (instead of doing query...first(), I just do the query and the function handles the rest when needed), and adds the result to that model's instance (since it's related) as a column that starts with "_". So `get_domain` goes in that instance dictionary as `_domain` = Domain's instance.

The next times the same function is called, since `_domain` is already present, it returns the value directly instead of calling the DB again.

I checked your fifth point. It's about `expire_on_commit` (If I counted properly) which is already set to False on my code. So this makes me wonder why my object expired then. Commit might have happened, but it shouldn't have expired in that case.

I'll read the last link soon.

**Reply by zzzeek** [reply]
if you have ORM objects that are inside of long-term memoized methods across sessions, that wont work.  objects that aren't detached are each bound to one session at at time and it needs to be the session that you're currently working with.     if an object bound to a session outlives the session, now it's detached.   you need to be clear about these states/boundaries in your program.  see https://docs.sqlalchemy.org/en/20/orm/session_state_management.html#quickie-intro-to-object-states

**Reply by cnicodeme** [reply]
Hi @zzzeek.

I had this thread open for quite some time (almost 2 years now). The difficulty for me here was to find a reproducible bug so that I could investigate and find the root cause of it.

Turns out I managed to push some code that made the bug reproducible every time, which led me to the cause (but not a solution yet, only ideas).

So to put simply, some of my model would fail with a 

> greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/20/xd2s)

When trying to access columns that were basic (String, Integers, etc). **Absolutely not a relationship or lazy loaded column**.

So, why ?

Being able to locate one key area that was causing the bug, I showed the attributes for the given model doing:

```python
from sqlalchemy.orm import attributes
attributes.instance_state(my_model_instance).dict
```

Before, and after the failing code.

Turns out, the model was fully loaded before, and after the failing code, the model had lost all attributes. (only `_sa_instance_state` was present).

Digging into that faulty code, I discovered that I had an `IntegrityError` triggered, properly handled in a `catch` block that contained a rollback there:

```python
try:
    await some_model.save()
catch IntegrityError:
    await session.rollback()
```

I thought this was enough, but it turns out that having an IntegrityError seems to clear all the models (in that session?) to be "emptied" (I'm sure there is an appropriate term for that in SQLAlchemy).

So then, when I would access a simple column in the original model, since it was now absent, SQLAlchemy would try to refresh the model, but since in async mode, it would fail because the current call would be without `await` ( a simple `my_model.some_column` in either get or set action).

---

Ok, so now, what could I do to avoid this ?

a) My first approach to solve this was to use a nested transaction, so that if it fails, it would fail under that current transaction and not impact the parent's transaction. It turns out that if there is an IntegrityError, the whole "session tree" is impacted, so this solution doesn't work.

b) I couldn't also create a new separate session for this, as it depends on the current session (models being created with their dedicated ID).

c) I couldn't call something like `await my_model.refresh()` right after to handle the case it would need to reload because I have many models so if I solve for this one, I would have to apply this code for all the loaded model when they are used (which is happening after an IntegrityError)

d) A potential solution (?) would be to eventually use SQLAlchemy to retrieve all the models in loaded in the current session that were "emptied" because of the IntegrityError, and reload them when an IntegrityError occurs. Doing that, I could reload them with a proper `await`. This makes sense, but **it doesn't look very efficient**.

@zzzeek Would you have a better recommandation here ? I'm already glad I found the root issue in my case as it was not related to relationships or lazy loading. IntegrityError is a new area that I hadn't thought about (maybe there are others).

I must say that so far, using `async` is bringing some headaches that would be completely absent in a non-asynchronous world, and I'm hesitant to switch back.

Thank you for your help here @zzzeek , sorry for the long (but hopefully clear and detailled) comment.

**Reply by CaselIT** [reply]
what you describe is a state that's called expired and it's documented here https://docs.sqlalchemy.org/en/20/orm/session_state_management.html#session-expire

You can use https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs to access attributes asynchronously

**Reply by cnicodeme** [reply]
@CaselIT Thank you, that makes sense, but unfortunately my issue remains.

The problem is that I can't anticipate from a section of my code if another place will trigger an IntegrityError such as:

```
... in my code A loading my model into the variable my_model

... calling code B
... code B triggers an IntegrityError, which is handled and ignored (A table with constraint, if it is already present, no need to add it again, which is the reason for the integrity error)
... code B finishes and return to code A

... code A tries to access a value at my_model.some_value
... > greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place?
```

In that scenario, it makes no sense to reload every model after I have called "code B" just in case it failed.

So far, what I'm trying to do is when in a code place that could trigger an `IntegrityError`, I would:

Save the current state of any modified models
Run the code
if fails, refresh each instance and re-set the modified elements

But again, I'm not sure this is the best approach...

**Reply by cnicodeme** [reply]
I came up with this:

```python
from sqlalchemy import inspect
from sqlalchemy.orm import attributes


class IntegrityGuardContext:
    def __init__(self, dbsession):
        self.dbsession = dbsession
        self.changes = {}

    async def __aenter__(self):
        # We save the changes
        for (key, instance) in list(self.dbsession.identity_map.items()):
            state = inspect(instance)
            if state.modified:
                self.changes[key] = {attr.key: state.attrs[attr.key].value for attr in state.attrs if state.attrs[attr.key].history.has_changes()}

        return self.dbsession

    async def __aexit__(self, exc_type, exc_value, traceback):
        if exc_type is None:
            return

        if issubclass(exc_type, exc.IntegrityError):
            await self.dbsession.rollback()
            for (key, instance) in list(self.dbsession.identity_map.items()):
                await db.write.refresh(instance)
                if key in self.changes:
                    for attr in self.changes[key]:
                        setattr(instance, attr, self.changes[key][attr])

            return True

        # Otherwise, we raise
        raise exc_value
```

And so, around the code that might trigger an IntegrityError I want to ignore:

```python
async with IntegrityGuardContext(session):
    await my_model.save()
```

That's all. If no IntegrityError is triggered, it will continue working as usual. If an IntegrityError is triggered, it will refresh all the instances and add back all the tracked changes.

Is there something wrong with that?

(My only bother, is the potential size that `self.changes` can grow to)

**Reply by CaselIT** [reply]
the current logic is like that since after a rollback you cannot assume that the view of the database that you have loaded in the object is still consistent.

I think the correct approach would be to avoid integrity error, since that's an unexpected state, it shouldn't be the "normal" flow of the code.

> Thank you, that makes sense, but unfortunately my issue remains.

I don't see how your issue remains if you use async_attrs. It will allow you to load expired data on demand, like when using sync code

**Reply by cnicodeme** [reply]
Hi @CaselIT, thank you for your reply (and sorry for my late response here).

The logic of considering all objects to be stale after a rollback is rather opiniated; The whole "loading data ; trying some insert/update; having an IntegrityError; Doing a rollback" is generally done in a few milliseconds.

Your suggestion to avoid IntegrityError can be impossible in some cases, for instance when multiple workers are working on data in the same table. Doing a verification before inserting is also tricky as there can be a race condition issue.

I think IntegrityError are inevitable, and in this scenario, it would require reloading all the objects that were expired from the database even though they are still valid.

So I was wondering if there was an option just like `expire_on_commit` for rollbacks ?

Also, do you know in the event of a rollback, if it's all the loaded objects from the current session that are expired, or only the failed object (the one that resulted in an IntegrityError).

Thank you :)

**Reply by zzzeek** [reply]
>  It turns out that if there is an IntegrityError, the whole "session tree" is impacted, so this solution doesn't work.

the correct way to catch IntegrityError and keep going is to use nested transactions e.g. SAVEPOINT.  that's the only pattern we offer.

I dont know what "the whole session tree is impacted" means.

**Reply by zzzeek** [reply]
> So I was wondering if there was an option just like expire_on_commit for rollbacks ?

the problem is the Session has accumulated all kinds of state about what objects need to yet be inserted, deleted, or updated, and when the DB has a rollback, none of that information is reliable anymore.  the Session does not keep a log of every flush operation and everything that changed in all of those operations.    auto-generated primary key values are all toast, the foreign keys that refer to them are toast, etc.   leaving any of that state around means things will break on the next operation.

**Reply by cnicodeme** [reply]
@zzzeek By session tree, I meant all the objects that were loaded inside that transaction (I agree it wasn't clear, my bad).

Ok, the nested transaction seems like a good option, I'll see to implement it on my code.

Thank you :)

**Reply by zzzeek** [reply]
the nested transaction thing is designed to only impact the objects that you have modified within the scope of that transaction.  so if you have a specific object that you want to catch IntegrityError, only work with that one object inside the nested transaction block.  that object will be the only one expired if there's an error with a rollback-to-savepoint, everything else should remain unaffected.

this is a big feature inside the session that has a lot of design behind it

**Reply by cnicodeme** [reply]
That's interesting! Thinking about it, it does seems to be quit complicated to handle expiration only for those inside the nested transaction. I'll dig more on this area. Thank you !
